In [25]:
import pandas as pd

# Specify relative path to the file
file_path = "../database/CreditCard.csv"

# Load data into a pandas DataFrame
df = pd.read_csv(file_path)



# Descrição dos Dados
7043 observações com 33 variáveis

**CustomerID:** Um ID único que identifica cada cliente.

**Count:** Um valor usado em relatórios/painéis para somar o número de clientes em um conjunto filtrado.

**Country:** O país de residência principal do cliente.

**State:** O estado de residência principal do cliente.

**City:** A cidade de residência principal do cliente.

**Zip Code:** O código postal da residência principal do cliente.

**Lat Long:** A latitude e longitude combinadas da residência principal do cliente.

**Latitude:** A latitude da residência principal do cliente.

**Longitude:** A longitude da residência principal do cliente.

**Gender:** O gênero do cliente: Masculino, Feminino

**Senior Citizen:** Indica se o cliente tem 65 anos ou mais: Sim, Não

**Partner:** Indica se o cliente tem um parceiro: Sim, Não

**Dependents:** Indica se o cliente vive com quaisquer dependentes: Sim, Não. Dependentes podem ser filhos, pais, avós, etc.

**Tenure Months:** Indica a quantidade total de meses que o cliente esteve com a empresa até o final do trimestre especificado acima.

**Phone Service:** Indica se o cliente assina um serviço de telefone residencial com a empresa: Sim, Não

**Multiple Lines:** Indica se o cliente assina várias linhas telefônicas com a empresa: Sim, Não

**Internet Service:** Indica se o cliente assina um serviço de Internet com a empresa: Não, DSL, Fibra Óptica, Cabo.

**Online Security:** Indica se o cliente assina um serviço de segurança online adicional fornecido pela empresa: Sim, Não

**Online Backup:** Indica se o cliente assina um serviço de backup online adicional fornecido pela empresa: Sim, Não

**Device Protection:** Indica se o cliente assina um plano adicional de proteção de equipamentos para seus equipamentos de Internet fornecidos pela empresa: Sim, Não

**Tech Support:** Indica se o cliente assina um plano adicional de suporte técnico da empresa com tempos de espera reduzidos: Sim, Não

**Streaming TV:** Indica se o cliente usa seu serviço de Internet para transmitir programação de televisão de um provedor terceirizado: Sim, Não. A empresa não cobra uma taxa adicional por este serviço.

**Streaming Movies:** Indica se o cliente usa seu serviço de Internet para transmitir filmes de um provedor terceirizado: Sim, Não. A empresa não cobra uma taxa adicional por este serviço.

**Contract:** Indica o tipo de contrato atual do cliente: Mês a Mês, Um Ano, Dois Anos.

**Paperless Billing:** Indica se o cliente escolheu a faturação sem papel: Sim, Não

**Payment Method:** Indica como o cliente paga a sua conta: Débito Bancário, Cartão de Crédito, Cheque Enviado por Correio

**Monthly Charge:** Indica a cobrança mensal total atual do cliente por todos os seus serviços da empresa.

**Total Charges:** Indica as cobranças totais do cliente, calculadas até o final do trimestre especificado acima.

**Churn Label:** Sim = o cliente deixou a empresa este trimestre. Não = o cliente permaneceu com a empresa. Diretamente relacionado ao Valor de Churn.

**Churn Value:** 1 = o cliente deixou a empresa este trimestre. 0 = o cliente permaneceu com a empresa. Diretamente relacionado ao Rótulo de Churn.

**Churn Score:** Um valor de 0-100 que é calculado usando a ferramenta preditiva IBM SPSS Modeler. O modelo incorpora múltiplos fatores conhecidos por causar churn. Quanto maior a pontuação, mais provável é que o cliente deixe a empresa.

**CLTV:** Valor do Ciclo de Vida do Cliente. Um CLTV previsto é calculado usando fórmulas corporativas e dados existentes. Quanto maior o valor, mais valioso é o cliente. Clientes de alto valor devem ser monitorados para churn.

**Churn Reason:** A razão específica do cliente para deixar a empresa. Diretamente relacionado à Categoria de Churn.



In [ ]:
df = df.drop_duplicates()

Dropar as duplicadas antes de removermos a ID individual que identifica cada individualmente cada cliente

In [26]:
df = df.drop(columns=['CustomerID'])

In [27]:
single_value_columns = [col for col in df.columns if df[col].nunique() <= 1]
df = df.drop(columns=single_value_columns)
print(f"As colunas {single_value_columns} foram removidas pois continham apenas um valor diferente e não acrescentavam nada à análise.")

As colunas ['Count', 'Country', 'State'] foram removidas pois continham apenas um valor diferente e não acrescentavam nada à análise.


In [28]:
df

City  Zip Code                Lat Long   Latitude   Longitude  \
0      Los Angeles     90003  33.964131, -118.272783  33.964131 -118.272783   
1      Los Angeles     90005   34.059281, -118.30742  34.059281 -118.307420   
2      Los Angeles     90006  34.048013, -118.293953  34.048013 -118.293953   
3      Los Angeles     90010  34.062125, -118.315709  34.062125 -118.315709   
4      Los Angeles     90015  34.039224, -118.266293  34.039224 -118.266293   
...            ...       ...                     ...        ...         ...   
7038       Landers     92285  34.341737, -116.539416  34.341737 -116.539416   
7039      Adelanto     92301  34.667815, -117.536183  34.667815 -117.536183   
7040         Amboy     92304  34.559882, -115.637164  34.559882 -115.637164   
7041  Angelus Oaks     92305     34.1678, -116.86433  34.167800 -116.864330   
7042  Apple Valley     92308  34.424926, -117.184503  34.424926 -117.184503   

      Gender Senior Citizen Partner Dependents  Tenure Months  ...  \
0       Male             No      No         No              2  ...   
1     Female             No      No        Yes              2  ...   
2     Female             No      No        Yes              8  ...   
3     Female             No     Yes        Yes             28  ...   
4       Male             No      No        Yes             49  ...   
...      ...            ...     ...        ...            ...  ...   
7038  Female             No      No         No             72  ...   
7039    Male             No     Yes        Yes             24  ...   
7040  Female             No     Yes        Yes             72  ...   
7041  Female             No     Yes        Yes             11  ...   
7042    Male             No      No         No             66  ...   

            Contract Paperless Billing             Payment Method  \
0     Month-to-month               Yes               Mailed check   
1     Month-to-month               Yes           Electronic check   
2     Month-to-month               Yes           Electronic check   
3     Month-to-month               Yes           Electronic check   
4     Month-to-month               Yes  Bank transfer (automatic)   
...              ...               ...                        ...   
7038        Two year               Yes  Bank transfer (automatic)   
7039        One year               Yes               Mailed check   
7040        One year               Yes    Credit card (automatic)   
7041  Month-to-month               Yes           Electronic check   
7042        Two year               Yes  Bank transfer (automatic)   

     Monthly Charges Total Charges Churn Label Churn Value Churn Score  CLTV  \
0              53.85        108.15         Yes           1          86  3239   
1              70.70        151.65         Yes           1          67  2701   
2              99.65         820.5         Yes           1          86  5372   
3             104.80       3046.05         Yes           1          84  5003   
4             103.70        5036.3         Yes           1          89  5340   
...              ...           ...         ...         ...         ...   ...   
7038           21.15        1419.4          No           0          45  5306   
7039           84.80        1990.5          No           0          59  2140   
7040          103.20        7362.9          No           0          71  5560   
7041           29.60        346.45          No           0          59  2793   
7042          105.65        6844.5          No           0          38  5097   

                       Churn Reason  
0      Competitor made better offer  
1                             Moved  
2                             Moved  
3                             Moved  
4     Competitor had better devices  
...                             ...  
7038                            NaN  
7039                            NaN  
7040                            NaN  
7041                            NaN  
7042            

# LIMPEZA DOS DADOS 

Checar se variáveis categóricas possuem valores não nulos, mas que são, na prática, valores nulos como "Unknown", "", "-", etc. 

In [29]:
columns = ['City', 'Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Phone Service', 'Multiple Lines', 
           'Internet Service', 'Online Security', 'Online Backup', 'Device Protection', 'Tech Support', 
           'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method', 
           'Churn Label', 'Churn Reason']

for col in columns:
    print(f"{col}: {df[col].unique()}")


City: ['Los Angeles' 'Beverly Hills' 'Huntington Park' ... 'Standish' 'Tulelake'
 'Olympic Valley']
Gender: ['Male' 'Female']
Senior Citizen: ['No' 'Yes']
Partner: ['No' 'Yes']
Dependents: ['No' 'Yes']
Phone Service: ['Yes' 'No']
Multiple Lines: ['No' 'Yes' 'No phone service']
Internet Service: ['DSL' 'Fiber optic' 'No']
Online Security: ['Yes' 'No' 'No internet service']
Online Backup: ['Yes' 'No' 'No internet service']
Device Protection: ['No' 'Yes' 'No internet service']
Tech Support: ['No' 'Yes' 'No internet service']
Streaming TV: ['No' 'Yes' 'No internet service']
Streaming Movies: ['No' 'Yes' 'No internet service']
Contract: ['Month-to-month' 'Two year' 'One year']
Paperless Billing: ['Yes' 'No']
Payment Method: ['Mailed check' 'Electronic check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn Label: ['Yes' 'No']
Churn Reason: ['Competitor made better offer' 'Moved' 'Competitor had better devices'
 'Competitor offered higher download speeds'
 'Competitor offered mo

In [30]:
null_values = df.isnull().sum()

# Imprimir o resultado
for col, null_count in null_values.iteritems():
    print(f"{col}: {null_count}")


City: 0
Zip Code: 0
Lat Long: 0
Latitude: 0
Longitude: 0
Gender: 0
Senior Citizen: 0
Partner: 0
Dependents: 0
Tenure Months: 0
Phone Service: 0
Multiple Lines: 0
Internet Service: 0
Online Security: 0
Online Backup: 0
Device Protection: 0
Tech Support: 0
Streaming TV: 0
Streaming Movies: 0
Contract: 0
Paperless Billing: 0
Payment Method: 0
Monthly Charges: 0
Total Charges: 0
Churn Label: 0
Churn Value: 0
Churn Score: 0
CLTV: 0
Churn Reason: 5174


Vemos que não há valores nulos, exceto em Churn Reason (Razões de cancelamento do cartão) em que existem 5174! Como nem todas as análises que faremos fará uso de Churn Reason, inicialmente não removerei os valores nulos desta variável. 

In [34]:
# Save cleaned DataFrame to a new CSV file
df.to_csv("../database/CreditCard_CLEANED.csv", index=False)
